changeset/changeset.go

```diff
@@ -9,88 +10,101 @@ import (
-	Touched    map[string]bool // map of field names that were touched
-	Values     url.Values      // map of merged changes and original values
-	Struct     any             // type of object to decode into
-	StructType string          // type name of the struct pointer
-
-	config *Config
+	Errors  map[string]error // map of field name to error message
+	Changes map[string]any   // map of field name that differs from the original value
+	Values  url.Values       // map of merged changes and original values
+	Struct  any              // pointer to struct to decode into
+
+	action  string          // last update action; only run validations if action is not empty
+	touched map[string]bool // map of field names that were touched
+	config  *Config
+}
+
+// Type returns the name of the struct.
+func (c *Changeset) Type() string {
+	return reflect.TypeOf(c.Struct).Elem().Name()
+}
+
+// Valid returns true if the changeset is valid or false if it is not.
+// Valid depends on the last action that was performed on the changeset along
+// with the Errors map and whether or not the field was touched.
+// If the action is empty, Valid will always return true.
+// If the action is not empty, Valid will return true if there are no errors
+// or if there are errors but the field was not touched.
+func (c *Changeset) Valid() bool {
+	if c.action == "" {
+		return true
+	}
+	for k, v := range c.Errors {
+		if v != nil {
+			if c.touched != nil {
+				return !c.touched[k]
+			}
+		}
+	}
+	return true
 }
 
-// AsStruct returns the changeset as a struct or an error if the data could not be decoded into the struct.
+// AsStruct returns the changeset decoded into a struct or an error if there was a problem decoding.
 func (c *Changeset) AsStruct() (any, error) {
 	s := c.Struct
-	err := c.config.decoder.Decode(s, c.Values)
+	err := c.config.Decoder.Decode(s, c.Values)
 	return s, err
 }
 
-// NewChangeset returns a new Changeset based on the old data, new data, and action
-// Typically this is called to initialize a changeset. If action is empty, the changeset
-// will always return true for Valid. Passing a non-empty action will cause the
-// changeset to make the validation errors available if the struct is not valid.
-func (cc *Config) NewChangeset(old, new url.Values, action string, obj any) *Changeset {
-	c := &Changeset{
-		Action:     action,
-		Valid:      action == "",
-		Errors:     make(map[string]any),
-		Changes:    make(map[string]any),
-		Touched:    make(map[string]bool),
-		Values:     url.Values{},
-		Struct:     obj,                               // TODO check is pointer to struct
-		StructType: reflect.TypeOf(obj).Elem().Name(), // TODO better way?
-		config:     cc,
+// NewChangeset returns a new Changeset based on the provided pointer to a struct.
+func (cc *Config) NewChangeset(obj any) (*Changeset, error) {
+	// we need a pointer to a struct to decode into
+	if reflect.TypeOf(obj).Kind() != reflect.Ptr {
```

Comment:
This checks that it is a pointer, but not that it is a pointer to a struct. Add the second check?